### ISSUE TITLE
Create a boxed fallback / template recipe for Autograd that forwards, but errors on backwards

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#62032]
FB only reference: https://fb.workplace.com/groups/pytorch.dev/posts/933475730564241/

There is a longstanding problem with custom operators to PyTorch, which is that they typically have incorrect autograd functionality. The most common practice (as documented in https://pytorch.org/tutorials/advanced/dispatcher.html) is if you write a forward-only operator, you simply omit the Autograd dispatch key definition. However, the default implementation of autograd in aten/src/ATen/core/VariableFallbackKernel.cpp simply *falls through*, so if any inputs require grad, this is silently lost. This can be **very** difficult to debug (that is what the FB reference above cites.) The correct behavior in this situation is to perform the forwards, but register the grad_fn NotImplemented (in torch/csrc/autograd/functions/basic_ops.h) which will report an error saying gradient is not implemented on backwards.

The first step to fixing this situation is making it possible to conveniently do this, in a single line. It is possible to do this by using the custom C++ autograd function interface, but it is very wordy: https://github.com/pytorch/vision/blob/b8a8fbc9933de1df3b269ac723e0dc66eefa4d80/torchvision/csrc/ops/autograd/roi_pool_kernel.cpp#L65-L102 There are two ways we can condense this:

1. Create a template function that generates all of the necessary boilerplate. As PyTorch core, we can do better than the sample code in torchvision; we should generate the equivalent of not implemented VariableType definitions (NB: I have excised the NDEBUG statements, but in a fully automated luxury space communism solution those would be implemented as well):

```
at::Tensor _add_relu_Tensor(c10::DispatchKeySet ks, const at::Tensor & self, const at::Tensor & other, const at::Scalar & alpha) {
  auto& self_ = unpack(self, "self", 0);
  auto& other_ = unpack(other, "other", 1);
  auto _any_requires_grad = compute_requires_grad( self, other );
  (void)_any_requires_grad;
  std::shared_ptr<NotImplemented> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<NotImplemented>(new NotImplemented("_add_relu"), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self, other ));
  }
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::_add_relu(ks & c10::after_autograd_keyset, self_, other_, alpha);
  })();
  auto result = std::move(_tmp);
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }   
  throw_error_for_complex_autograd(result, "_add_relu");
  TORCH_CHECK_NOT_IMPLEMENTED(!(isFwGradDefined(self) || isFwGradDefined(other)), "Trying to use forward AD with _add_relu that does not support it.");
  return result;
} 
```

This solution would be sufficient to let individual operator writers explicitly turn on correct autograd handling for their operators; but they would be required to do so explicitly. (It might be possible to modify def/impl to also automatically apply this registration when someone registers a CPU/CUDA implementation directly, but we would need to implement https://github.com/pytorch/pytorch/issues/35041 so that subsequent registrations of autograd would override this behavior.)

2. A simpler, albeit less efficient, way to solve this problem is to write the template function as a *boxed fallback*. We would thus instead write this function:

```
void autogradNotSupportedFallback(const c10::OperatorHandle& op, DispatchKeySet dispatch_keys, torch::jit::Stack* stack) {
...
}
```

This function would iterate over the stack checking for requires_grad'ness, and then set history for each of the outputs. There are many examples of existing boxed fallbacks and thus the implementation here should be relatively straightforward. The boxed fallback has an added benefit that we could then consider, at a later point in time, to register it as the universal Autograd fallback key, thus automatically marking all custom operators as not supporting autograd unless they explicitly register an Autograd implementation (this would be quite BC breaking and there would be quite a bit of fallout to clean up afterwards).

----

An orthogonal solution to this class of problems is to take @bdhirsh's work on external codegen and apply it to custom operators; in such situations, we could automatically codegen autograd kernels for custom operators. However, this is a bit further out, and I think the boxed fallback would be quickest to execute on.

**Comment by albanD** [comment]
Note that a related discussion I was having is that we should have a better differentiation between formulas that are not implemented compared to functions that are not differentiable.
While the two exists in derivatives.yaml today, the implementation is not very consistent and very clear about what each means.

I think that library writers might get into that same "dilemma" where not implemented should be the default, but they should be able to state that an op is just non-differentiable (and there doesn't need to be any autograd kernel).

Note that I would claim that we would want the default in core to be changed: everything is not-implemented unless specified otherwise. That will avoid issues with internal function that are in core.

**Comment by soulitzer** [comment]
Making a status update on the progress of this after prototyping in #62450 and offline discussion with @albanD and @ezyang - 

Prototyping and benchmarking helped us answer the following questions concerning some implementations details independent of the API design (a next step):

**Question 1: Do we want to have both boxed and templated kernel, or both?**
For boxed kernel:
 - slightly better UX  since with a templated kernel the user would need to provide a schema string
 - more benefits if we replace codegened kernels with boxed kernel (see below)

For templated kernel:
 - faster (we expected it to be, but turned out to be not very much faster <100ns difference [0])

Proposed answer to (1):  **boxed kernel only** because we avoid the extra burden of maintaining both, and there is not too much perf benefit as measured above.

**Question 2: Given our answer to (1), is desirable to replace codegened VariableType kernels with boxed/templated kernel**
Benefits of replacing:
 - reduced compilation time (16% reduction) [0]
 - smaller binaries (not yet measured)

Cost of replacing:
 - slightly slower (<100ns difference: [1])
 - must now maintain logic in both codegen  and boxed variabletype kernel

Proposed answer to (2): **Should replace** - there appears to be little difference between the two (<100ns). Maintain two sets of logic (between codegened and boxed kernel) should be OK as long as we cross reference them.

####  Steps to land the above (this will roughly correspond to stack of PRs):
1. Implement boxed kernel and testing
   - TODO: is there a way to land a test for this that would run on master without replacing the actual registrations
2. Replace codegened `VariableType` kernel for `NotImplemented` ops with our boxed kernel
3. User-facing API: an easy way to register this boxed fallback as autograd kernel
   - For now it should not be default since that would be pretty BC-breaking. 
   - In the future, we may automatically register `InplaceOrView` kernel IF user is registering a view and in-place op, but that should be done transparently and not affect this API.


[0] Benchmarking boxed vs templated VariableType kernel  (https://github.com/pytorch/pytorch/pull/62450)
[1] Benchmarking codegen'd vs boxed VariableType kernel (https://github.com/pytorch/pytorch/pull/63362)
<details>
<summary>
-O3
 </summary>

```
[<torch.utils.benchmark.utils.common.Measurement object at 0x7f672e294130>
torch::_test_op_for_benchmarking(a, b);
setup:
  torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
  torch::Tensor b = torch::ones({1, 2, 3}).set_requires_grad(false);

  Median: 1.73 us
  IQR:    0.02 us (1.72 to 1.74)
  289 measurements, 10000 runs per measurement, 1 thread] 

[<torch.utils.benchmark.utils.common.Measurement object at 0x7f65c566ff40>
torch::_test_op_for_benchmarking2(a, b);
setup:
  torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
  torch::Tensor b = torch::ones({1, 2, 3}).set_requires_grad(false);

  Median: 1.67 us
  IQR:    0.01 us (1.67 to 1.68)
  299 measurements, 10000 runs per measurement, 1 thread] 

51590
51786 

<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f672e301cd0>
    4070  ???:torch::autograd::autogradNotImplementedFallback(c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)
     480  ???:std::vector<bool, std::allocator<bool> >::_M_reallocate(unsigned long)
     470  ???:at::_ops::_test_op_for_benchmarking2::call(at::Tensor const&, at::Tensor const&)
     460  ???:std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > torch::autogr ... r<at::Tensor const*> >&>(std::vector<at::Tensor const*, std::allocator<at::Tensor const*> >&)
     325  ???:c10::impl::BoxedKernelWrapper<at::Tensor (at::Tensor const&, at::Tensor const&), void>:: ... rnel*, c10::OperatorHandle const&, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
     320  ???:void std::vector<c10::IValue, std::allocator<c10::IValue> >::emplace_back<at::Tensor const&>(at::Tensor const&)
     301  /build/glibc-eX1tMB/glibc-2.31/malloc/malloc.c:_int_free
     280  ???:void std::vector<at::Tensor const*, std::allocator<at::Tensor const*> >::_M_realloc_inse ... *, std::vector<at::Tensor const*, std::allocator<at::Tensor const*> > >, at::Tensor const*&&)
     265  ???:std::vector<c10::IValue, std::allocator<c10::IValue> >::_M_erase(__gnu_cxx::__normal_ite ... xx::__normal_iterator<c10::IValue*, std::vector<c10::IValue, std::allocator<c10::IValue> > >)
     ...
    -280  ???:at::_ops::_test_op_for_benchmarking::redispatch(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
    -285  ???:torch::autograd::set_history(at::Tensor&, std::shared_ptr<torch::autograd::Node> const&)
    -290  ???:torch::autograd::Node::set_next_edges(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >&&)
    -300  ???:std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > torch::autograd::collect_next_edges<at::Tensor const&, at::Tensor const&>(at::Tensor const&, at::Tensor const&)
    -415  ???:void c10::SmallVectorImpl<torch::autograd::InputMetadata>::emplace_back<at::Tensor const&>(at::Tensor const&)
    -420  ???:at::_ops::_test_op_for_benchmarking::call(at::Tensor const&, at::Tensor const&)
    -470  ???:at::Tensor::options() const
    -775  ???:void std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >::emplace_back<torch::autograd::Edge>(torch::autograd::Edge&&)
   -1370  ???:torch::autograd::VariableType::(anonymous namespace)::_test_op_for_benchmarking(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
  ```

</details>

<details>
<summary>
-O2 -g 
 </summary>

```
[<torch.utils.benchmark.utils.common.Measurement object at 0x7fd7b73058e0>
torch::_test_op_for_benchmarking(a, b);
setup:
  torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
  torch::Tensor b = torch::ones({1, 2, 3}).set_requires_grad(false);

  Median: 1.65 us
  IQR:    0.01 us (1.64 to 1.66)
  304 measurements, 10000 runs per measurement, 1 thread] 

[<torch.utils.benchmark.utils.common.Measurement object at 0x7fd64eb92e20>
torch::_test_op_for_benchmarking2(a, b);
setup:
  torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
  torch::Tensor b = torch::ones({1, 2, 3}).set_requires_grad(false);

  Median: 1.64 us
  IQR:    0.01 us (1.63 to 1.65)
  305 measurements, 10000 runs per measurement, 1 thread] 

51572
51702 

<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fd64f830250>
    805  build/../torch/csrc/autograd/custom_function.cpp:torch::autograd::autogradNotImplementedFallb ... rHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)
    450  build/../c10/core/TensorOptions.h:torch::autograd::autogradNotImplementedFallback(c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)
    450  /usr/include/c++/9/bits/stl_vector.h:torch::autograd::autogradNotImplementedFallback(c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)
    395  /usr/include/c++/9/bits/stl_bvector.h:torch::autograd::autogradNotImplementedFallback(c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)
    305  /build/glibc-eX1tMB/glibc-2.31/malloc/malloc.c:_int_free
    270  /usr/include/c++/9/bits/basic_string.h:torch::autograd::autogradNotImplementedFallback(c10::O ... rHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)
    260  /usr/include/c++/9/bits/vector.tcc:std::vector<bool, std::allocator<bool> >::_M_reallocate(unsigned long)
    230  /usr/include/c++/9/bits/vector.tcc:void std::vector<c10::IValue, std::allocator<c10::IValue> >::emplace_back<at::Tensor const&>(at::Tensor const&)
    200  build/../aten/src/ATen/core/boxing/impl/boxing.h:c10::impl::BoxedKernelWrapper<at::Tensor (at ... rnel*, c10::OperatorHandle const&, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
    ...
   -180  /usr/include/c++/9/bits/stl_vector.h:void std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >::emplace_back<torch::autograd::Edge>(torch::autograd::Edge&&)
   -190  build/../torch/csrc/autograd/function.h:std::vector<torch::autograd::Edge, std::allocator<tor ... ollect_next_edges<at::Tensor const&, at::Tensor const&>(at::Tensor const&, at::Tensor const&)
   -205  build/../torch/csrc/autograd/function.h:torch::autograd::Node::set_next_edges(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >&&)
   -210  build/../torch/csrc/autograd/VariableTypeManual.cpp:torch::autograd::VariableType::(anonymous namespace)::checked_cast_variable(at::Tensor const&, char const*, int)
   -215  build/../c10/util/SmallVector.h:void c10::SmallVectorImpl<torch::autograd::InputMetadata>::emplace_back<at::Tensor const&>(at::Tensor const&)
   -250  /usr/include/c++/9/bits/vector.tcc:void std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >::emplace_back<torch::autograd::Edge>(torch::autograd::Edge&&)
   -300  /usr/include/c++/9/bits/basic_string.h:torch::autograd::VariableType::(anonymous namespace)::_test_op_for_benchmarking(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
   -305  build/../c10/core/TensorOptions.h:at::Tensor::options() const
   -380  build/../torch/csrc/autograd/generated/VariableType_2.cpp:torch::autograd::VariableType::(ano ... espace)::_test_op_for_benchmarking(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)

Total: 130
```

</details>

**Comment by ezyang** [comment]
okey dokey

**Comment by soulitzer** [comment]
Step 3 of the above plan is blocked on the creation of a InplaceOrView kernel, which we need if we allow users to implement view and in-place custom ops. We proceed with landing InplaceOrView kernel in two steps:

1. Just like the autograd not implemented fallback, the InplaceOrView kernel could either be boxed or unboxed. Ideally this InplaceOrView kernel will be boxed because in the templated (unboxed) case users would need to pass in a schema string. We should benchmark prototypes of both, and confirm that boxed is not significantly slower than a templated version.

2. If we decide to proceed with the boxed kernel we can also consider replacing the codegened InplaceOrView kernels with a boxed kernel. This has the benefit of reducing compilation time and binary size. Just like for autograd fallback, we should benchmark both codegened and boxed kernel to make sure there isn't a major regression.

**Comment by soulitzer** [comment]
Update:
 - creating a new issue to track the addition of ADInplaceOrView https://github.com/pytorch/pytorch/issues/64407
 - Re-ran benchmarking of the boxed autograd not implemented kernel, to provide percentages. Also ran a better baseline for comparison. (These are the same numbers as expected, but the percentages may seem a little alarming - either way, it is easy to toggle codegen to use/not use this new kernel)

Benchmarking:
 - Using boxed kernel over codegened VariableType kernel increases instruction count by 7% (700 instructions) for
a dummy non-inplace, non-view op (that just does addition). 

<details>
<summary>
details
</summary>

```

[<torch.utils.benchmark.utils.common.Measurement object at 0x7fbec9f670d0>
torch::my_test_normal_op_boxed(a, b);
setup:
  auto a = torch::ones({1, 2, 3}).set_requires_grad(true).clone();
  auto b = torch::ones({1, 2, 3}).set_requires_grad(true).clone();

  Median: 1.62 us
  IQR:    0.01 us (1.62 to 1.63)
  308 measurements, 10000 runs per measurement, 1 thread] 

[<torch.utils.benchmark.utils.common.Measurement object at 0x7fbd6106f430>
torch::my_test_normal_op(a, b);
setup:
  auto a = torch::ones({1, 2, 3}).set_requires_grad(true).clone();
  auto b = torch::ones({1, 2, 3}).set_requires_grad(true).clone();

  Median: 1.56 us
  IQR:    0.02 us (1.55 to 1.58)
  320 measurements, 10000 runs per measurement, 1 thread] 

10381
9678 

<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fbec9f27f70>
    178  ???:torch::autograd::VariableType::(anonymous namespace)::my_test_normal_op(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
    155  ???:void std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >::emplace_back<torch::autograd::Edge>(torch::autograd::Edge&&)
     99  ???:torch::autograd::NotImplemented::NotImplemented(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
     84  ???:at::_ops::my_test_normal_op::call(at::Tensor const&, at::Tensor const&)
     62  ???:torch::autograd::Node::set_next_edges(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >&&)
     60  ???:std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > torch::autograd::collect_next_edges<at::Tensor const&, at::Tensor const&>(at::Tensor const&, at::Tensor const&)
     57  ???:torch::autograd::set_history(at::Tensor&, std::shared_ptr<torch::autograd::Node> const&)
     56  ???:torch::autograd::Node::Node(unsigned long, std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> >&&)
     56  ???:at::_ops::my_test_normal_op::redispatch(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
    ...
    -65  ???:c10::impl::BoxedKernelWrapper<at::Tensor (at::Tensor const&, at::Tensor const&), void>::c ... rnel*, c10::OperatorHandle const&, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
    -86  ???:torch::autograd::autogradNotImplementedFallbackImpl(c10::OperatorHandle const&, c10::Disp ... ue, std::allocator<c10::IValue> >*)::{lambda(unsigned long, unsigned long, at::Tensor const&)
    -94  ???:at::_ops::my_test_normal_op_boxed::call(at::Tensor const&, at::Tensor const&)
    -96  ???:std::vector<bool, std::allocator<bool> >::_M_reallocate(unsigned long)
   -118  /build/glibc-eX1tMB/glibc-2.31/malloc/malloc.c:malloc
   -131  ???:void std::vector<at::Tensor const*, std::allocator<at::Tensor const*> >::_M_realloc_inser ... *, std::vector<at::Tensor const*, std::allocator<at::Tensor const*> > >, at::Tensor const*&&)
   -187  /build/glibc-eX1tMB/glibc-2.31/malloc/malloc.c:_int_free
   -196  ???:std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > torch::autogra ... r<at::Tensor const*> >&>(std::vector<at::Tensor const*, std::allocator<at::Tensor const*> >&)
   -669  ???:torch::autograd::autogradNotImplementedFallbackImpl(c10::OperatorHandle const&, c10::DispatchKeySet, std::vector<c10::IValue, std::allocator<c10::IValue> >*)

Total: -703

```

</details>

 - The baseline (using inference mode to skip VariableType kernel) for non-inplace, non-view op uses 5252 instructions.
Subtracting this baseline from the above measurements, the instruction counts are 4426 (codegened) and 5129 (boxed), a 16%
increase.

<details>
<summary>
details
</summary>

```

[<torch.utils.benchmark.utils.common.Measurement object at 0x7fa592f8c0d0>
torch::my_test_normal_op_boxed(a, b);
setup:
  auto a = torch::ones({1, 2, 3}).set_requires_grad(true).clone();
  auto b = torch::ones({1, 2, 3}).set_requires_grad(true).clone();

  c10::InferenceMode guard;

  Median: 715.59 ns
  IQR:    18.66 ns (704.64 to 723.29)
  699 measurements, 10000 runs per measurement, 1 thread] 

[<torch.utils.benchmark.utils.common.Measurement object at 0x7fa42a089a30>
torch::my_test_normal_op(a, b);
setup:
  auto a = torch::ones({1, 2, 3}).set_requires_grad(true).clone();
  auto b = torch::ones({1, 2, 3}).set_requires_grad(true).clone();

  c10::InferenceMode guard;

  Median: 700.73 ns
  IQR:    8.02 ns (696.50 to 704.52)
  714 measurements, 10000 runs per measurement, 1 thread] 

5252
5252 

<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fa592f4cf70>
    84  ???:at::_ops::my_test_normal_op::call(at::Tensor const&, at::Tensor const&)
    25  /tmp/benchmark_utils_jit_build__1630419839__d9626b2b-10a3-4ba8-9541-3a08d5cb4c46/timer_cpp_4457552915623088790/timer_src.cpp:main
    18  ???:c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10:: ... onst&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
   -18  ???:c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10:: ... onst&)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&)
   -25  /tmp/benchmark_utils_jit_build__1630419839__d9626b2b-10a3-4ba8-9541-3a08d5cb4c46/timer_cpp_2736991312630238595/timer_src.cpp:main
   -84  ???:at::_ops::my_test_normal_op_boxed::call(at::Tensor const&, at::Tensor const&)

Total: 0

```

</details>

**To replicate:**
 - Branch: https://github.com/pytorch/pytorch/pull/64021
<details>
<summary>
code
</summary>

```python
# From https://github.com/pytorch/pytorch/issues/55126
import timeit
import torch
from torch.utils.benchmark import Timer

def make_timer_view(op_name, boxed):
    setup_two_tensors = """
    auto a = torch::ones({1, 2, 3}).set_requires_grad(true).clone();
    auto b = torch::ones({1, 2, 3}).set_requires_grad(true).clone();
    """
    if "view" in op_name:
        args = "a"
    else:
        args = "a, b"
    if "inplace" in op_name:
        suffix = "_"
    else:
        suffix = ""
    if boxed:
        return Timer(f"torch::{op_name}_boxed{suffix}({args});", setup=setup_two_tensors, language="c++", timer=timeit.default_timer)
    else:
        return Timer(f"torch::{op_name}{suffix}({args});", setup=setup_two_tensors, language="c++", timer=timeit.default_timer)

def measure_main(op_name, boxed):
    timer = make_timer_view(op_name, boxed)
    counts = timer.collect_callgrind(number=1, repeats=1, collect_baseline=False)
    times = timer.blocked_autorange(min_run_time=5)
    return counts, times

def main(**kwargs):
    for op_name in ["my_test_inplace_op"]:
        print(f'{"="*50}\n{op_name.center(50)}\n{"="*50}')
        times = [[], []]
        counts = [[], []]
        c, t = measure_main(op_name, True)
        times[0].append(t)
        counts[0].append(c)
        c, t = measure_main(op_name, False)
        times[1].append(t)
        counts[1].append(c)

        print()

        t0 = times[0]
        t1 = times[1]

        # Take the min as any interpreter jitter will increase from the baseline.
        c0 = min(counts[0][0], key=lambda x: x.counts(denoise=True))
        c1 = min(counts[1][0], key=lambda x: x.counts(denoise=True))

        torch.set_printoptions(linewidth=200)
        delta = (
            c1.as_standardized().stats() -
            c0.as_standardized().stats()
        ).denoise()

        print(t0, "\n")
        print(t1, "\n")
        print(c0.counts(denoise=True))
        print(c1.counts(denoise=True), "\n")
        print(delta)

if __name__ == "__main__":
    main()

```

</details>

**Comment by ezyang** [comment]
As expected, I imagine!
